README.md
## 研究開発用ロボットプラットフォーム Robovie-R4 ROSパッケージ

<p align="center">
  <img src="./images/RobovieR4_01.jpg" width="600" />
</p>

ヴイストン株式会社より発売されている開発研究用ロボットプラットフォーム「Robovie-R4」をROSで制御するためのSDKです。別途Linux搭載のPC及びロボット実機が必要になります。

Robovie-R4については、[製品ページ](https://www.vstone.co.jp/products/robovie_r4/index.html)をご覧ください。

## 目次
- [インストール方法](#インストール方法)
- [各パッケージの説明と利用方法](#各パッケージの説明と利用方法)
    - [r4\_control](#r4_control)
    - [r4\_description](#r4_description)
    - [r4\_motion](#r4_motion)
    - [r4\_moveit\_config](#r4_moveit_config)
      - [MoveIt](#moveit)
      - [Gazeboシミュレータ](#gazeboシミュレータ)
- [Robovie-R4の制御方法](#robovie-r4の制御方法)
  - [ROS（ROS1）メッセージ通信](#rosros1メッセージ通信)
      - [Subscribe](#subscribe)
      - [Publish](#publish)
  - [MoveIt＋RVizで動かす](#moveitrvizで動かす)

# インストール方法

- ROSのインストール方法は[こちら](http://wiki.ros.org/noetic/Installation/Ubuntu)を確認してください。
- catkin ワークスペースを作成
  ```bash
  mkdir -p ~/robovie_ws/src
  ```
- ワークスペースの`src`フォルダ内に移動し、本パッケージをクローンします。
  ```bash
  cd ~/robovie_ws/src
  git clone https://github.com/vstoneofficial/robovie_r4_ros.git
  ```
- Robovie-R4移動のための台車部分に研究開発用台車ロボットとして定評のある「[メガローバーVer.3.0](https://www.vstone.co.jp/products/wheelrobot/ver.3.0_normal.html)」を採用しています。\
  ROS上でRobovie-R4を動作したい場合、下記のパッケージが必要です。
  ```bash
  git clone https://github.com/vstoneofficial/megarover3_ros.git  # メガローバーVer.3.0のパッケージ
  git clone https://github.com/vstoneofficial/megarover_description.git  # メガローバーの3Dモデルのパッケージ 
  ```
- 依存関係にあるパッケージをインストールします。
  ```bash
  rosdep install -r -y --from-paths . --ignore-src
  ```
- ソースコードをビルドします。
  ```bash
  cd ~/robovie_ws
  catkin_make
  ```
- ワークスペースのオーバレイ作業
  ```bash
  source devel/setup.bash
  ```

以上で`robovie_r4_ros`パッケージのセットアップは完了です。

# 各パッケージの説明と利用方法
### r4_control
Robovie-R4をROS Controlで制御するパッケージ。\
以下のコマンドでRobovie-R4をROSと接続。
```
roslaunch r4_control bringup.launch
```

以下のコマンドでRobovie-R4をros_controlで制御する
```
roslaunch r4_control r4_control_HW.launch
```

実機を操作する前に、[RViz](http://wiki.ros.org/rviz)上で動作確認のためにこちらのノードを起動してください。

> **Warning**\
> 実機のRobovie-R4を操作する時に起動しないでください。

```
rosrun r4_control r4_sim_echo
```

### r4_description
Robovie-R4のモデルを表示するためのパッケージ。\
以下のコマンドで立ち上げます。
```
roslaunch r4_description display.launch
```
以下の画面が表示します。
![](images/r4-rviz.png)

### r4_motion
Robovie-R4のサンプルモーション（MoveItへ指令を出すノード）のパッケージ。\
各ノードの詳細は`src`フォルダーの各ファイルに記載されています。

### r4_moveit_config
Robovie-R4 MoveItの設定パッケージ
#### MoveIt
以下のコマンドでRViz上でMoveItを操作するデモを立ち上げてください。
```
roslaunch r4_moveit_config demo.launch
```
![](images/r4-moveit.png)

#### Gazeboシミュレータ
以下のコマンドでGazeboでRobovie-R4シミュレーションを起動します。
```
roslaunch r4_moveit_config gazebo.launch
```
MoveItで操作するGazeboのシミュレーションは次のコマンドで立ち上げてください。
```
roslaunch r4_moveit_config demo_gazebo.launch
```
> **Note**
> パラメータチューニングが必要な場合はあります。

# Robovie-R4の制御方法
## ROS（ROS1）メッセージ通信
Robovie-R4 の subscribe（購読）するメッセージ、publish（配信）するメッセージの仕様は以下のようになります。
#### Subscribe
- `/r4/r4Cmd`
  - 内容
      ```
      int16[11] angle # 各軸の目標角度の指令値 [0.1 deg]
      uint16[3] led # 目のLED（R, G, B）
      int16 time # 目標角度までの遷移時間 [16.667 msec]
      ```
      
#### Publish
- `/r4/r4Sensor`
  - 内容
      ```
      int16[11] angle # 各軸の現在角度 [0.1 deg]
      ```
      

## MoveIt＋RVizで動かす
上記の [r4\_control](#r4_control) のコマンドを実行してから、GUI上の操作が可能です。
MoveItの使い方の詳細は[公式のドキュメント](https://ros-planning.github.io/moveit_tutorials/index.html)を参照してください。


r4_motion/CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package r4_motion
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.0.1 (2023-03-16)
------------------
* docs: create CHANGELOGs
* Contributors: jie

1.0.0 (2023-03-15)
------------------
* 1.0.0
* Contributors: jie


r4_control/src/r4_hw_main.cpp
#include <ros_control_boilerplate/generic_hw_control_loop.h>
#include <r4_control/r4_hw_interface.h>

int main(int argc, char** argv)
{
  ros::init(argc, argv, "r4_hw_interface");
  ros::NodeHandle nh;

  // NOTE: We run the ROS loop in a separate thread as external calls such
  // as service callbacks to load controllers can block the (main) control loop
  ros::AsyncSpinner spinner(3);
  spinner.start();

  // Create the hardware interface specific to your robot
  std::shared_ptr<r4_control_ns::R4HWInterface> r4_hw_interface_instance(
      new r4_control_ns::R4HWInterface(nh));
  r4_hw_interface_instance->init();

  // Start the control loop
  ros_control_boilerplate::GenericHWControlLoop control_loop(nh, r4_hw_interface_instance);
  control_loop.run();  // Blocks until shutdown signal recieved

  return 0;
}

r4_control/src/r4_hw_interface1.cpp
#include <r4_control/r4_hw_interface.h>
#include <control_msgs/FollowJointTrajectoryActionGoal.h>
#include <std_msgs/Float32MultiArray.h>

namespace r4_control_ns
{
    R4HWInterface::R4HWInterface(ros::NodeHandle &nh, urdf::Model *urdf_model)
        : ros_control_boilerplate::GenericHWInterface(nh, urdf_model)
    {
        sensor_sub = nh.subscribe("/r4/r4Sensor", 1, &R4HWInterface::sensorCallback, this);
        goal_pos_sub_l_arm = nh.subscribe("/l_arm_position_trajectory_controller/follow_joint_trajectory/goal", 1, &R4HWInterface::lArmGoalCallback, this);
        goal_pos_sub_r_arm = nh.subscribe("/r_arm_position_trajectory_controller/follow_joint_trajectory/goal", 1, &R4HWInterface::rArmGoalCallback, this);
        goal_pos_sub_head = nh.subscribe("/head_position_trajectory_controller/follow_joint_trajectory/goal", 1, &R4HWInterface::headGoalCallback, this);

        cmd_pub = nh.advertise<r4_control::r4Cmd>("/r4/r4Cmd", 3);
        ROS_INFO("R4HWInterface declared.");
    }

    void R4HWInterface::lArmGoalCallback(const control_msgs::FollowJointTrajectoryActionGoal::ConstPtr &msg)
    {
        int waypoints = msg->goal.trajectory.points.size();
        // ROS_INFO("waypoints: %d", waypoints);
        // Assume the final point of trajectory as our goal, and set it to joint_position_command_
        // for (int i = 0; i < msg->goal.trajectory.points[waypoints - 1].positions.size(); i++)
        // pitch, roll, yaw, bend

        //bend, yaw, pitch, roll
        final_angle[0] = msg->goal.trajectory.points[waypoints - 1].positions[2]; 
        final_angle[1] = msg->goal.trajectory.points[waypoints - 1].positions[3];
        final_angle[2] = msg->goal.trajectory.points[waypoints - 1].positions[1];
        final_angle[3] = msg->goal.trajectory.points[waypoints - 1].positions[0];

        duration = msg->goal.trajectory.points[waypoints - 1].time_from_start.toSec();
    }

    void R4HWInterface::rArmGoalCallback(const control_msgs::FollowJointTrajectoryActionGoal::ConstPtr &msg)
    {
        int waypoints = msg->goal.trajectory.points.size();
        // ROS_INFO("waypoints: %d", waypoints);
        // Assume the final point of trajectory as our goal, and set it to joint_position_command_
        // for (int i = 0; i < msg->goal.trajectory.points[waypoints - 1].positions.size(); i++)
        // pitch, roll, yaw, bend

        //bend, yaw, pitch, roll
        final_angle[4] = msg->goal.trajectory.points[waypoints - 1].positions[2];
        final_angle[5] = msg->goal.trajectory.points[waypoints - 1].positions[3];
        final_angle[6] = msg->goal.trajectory.points[waypoints - 1].positions[1];
        final_angle[7] = msg->goal.trajectory.points[waypoints - 1].positions[0];
        
        duration = msg->goal.trajectory.points[waypoints - 1].time_from_start.toSec();
    }

    // want: yaw roll pitch
    // got: pitch roll yaw
    void R4HWInterface::headGoalCallback(const control_msgs::FollowJointTrajectoryActionGoal::ConstPtr &msg)
    {
        int waypoints = msg->goal.trajectory.points.size();
        // ROS_INFO("waypoints: %d", waypoints);
        // Assume the final point of trajectory as our goal, and set it to joint_position_command_
        // for (int i = 0; i < msg->goal.trajectory.points[waypoints - 1].positions.size(); i++)
        // pitch, roll, yaw, bend

        //bend, yaw, pitch, roll
        final_angle[8]  = msg->goal.trajectory.points[waypoints - 1].positions[2];
        final_angle[9]  = msg->goal.trajectory.points[waypoints - 1].positions[1];
        final_angle[10] = msg->goal.trajectory.points[waypoints - 1].positions[0];
        
        duration = msg->goal.trajectory.points[waypoints - 1].time_from_start.toSec();
    }

    void R4HWInterface::sensorCallback(const r4_control::r4Sensor::ConstPtr &msg)
    {
        for (int i = 0; i < num_joints_; i++)
        {
            joint_position_[i] = msg->angle[i] * DEG_TO_RAD / 10.0; // angle received is multiplied by 10
        }
    }

    void R4HWInterface::init()
    {
        // Call parent class version of this function
        ros_control_boilerplate::GenericHWInterface::init();

        final_angle.resize(joint_position_.size());
        joint_position_prev_.resize(joint_position_.size());
        joint_position_command_.resize(joint_position_.size());

        ROS_INFO("R4HWInterface initiated.");
        // ROS_INFO("num_joints_: %ld", num_joints_);
    }

    void R4HWInterface::read(ros::Duration &elapsed_time)
    {
        // No need to read since our write() command populates our state for us
        // ros::spinOnce();
    }

    void R4HWInterface::write(ros::Duration &elapsed_time)
    {
        static r4_control::r4Cmd joint_cmd;
        const double SAMPLE_RATE = 60.0; // ポーズ間の時間係数　higher value makes robot move faster

        bool change_detected = false;
        for (int i = 0; i < num_joints_; i++)
        {
            if (joint_position_prev_[i] != joint_position_command_[i])
            {
                change_detected = true;
                i = static_cast<int>(num_joints_);
            }
        }
        double percent_moved;
        if (change_detected)
        {
            for (int i = 0; i < num_joints_; i++)
            {
                joint_cmd.angle[i] = static_cast<int16_t>(joint_position_command_[i] * RAD_TO_DEG * 10);
                // joint_cmd.angle[i] = static_cast<int16_t>(joint_position_command_[i]*10000); // just for debugging
                // joint_cmd.angle[i] = static_cast<int16_t>(final_angle[i] * RAD_TO_DEG * 10);
                // joint_cmd.vel[i] = joint_velocity_command_[i] * RAD_TO_MRAD;

                joint_position_prev_[i] = joint_position_command_[i];
                // percent_moved = joint_position_command_[0] / final_angle[0];
            }

            // ROS_INFO("duration: %.4f", duration);

            joint_cmd.time = static_cast<int16_t>((duration * 1000.0) / 16.66666667 / SAMPLE_RATE);
            // joint_cmd.time = static_cast<int16_t>(percent_moved * 100);
            cmd_pub.publish(joint_cmd);
        }
    }

    void R4HWInterface::enforceLimits(ros::Duration &period)
    {
        // Enforces position and velocity
        // pos_jnt_sat_interface_.enforceLimits(period);
        // vel_jnt_sat_interface_.enforceLimits(period);
    }

} // namespace r4_control_ns


r4_motion/src/move_group_joint_value.cpp
/*
 * 各軸の目標角度を指定してRobovie-R4を動かすサンプル
 */

#include <ros/ros.h>
#include <moveit/move_group_interface/move_group_interface.h>
#include <moveit_visual_tools/moveit_visual_tools.h>

// 最大関節速度、加速度を下げるためのスケーリング係数の設定。許可される値は0～1です。
// ロボットモデルで指定された最大関節速度に係数が乗算されます。
const double MAX_VEL_ARM = 0.4;
const double MAX_ACC_ARM = 0.5;
const double MAX_VEL_HEAD = 1.0;
const double MAX_ACC_HEAD = 1.0;

void moveArm(moveit::planning_interface::MoveGroupInterface& group, std::vector<double>& angles, double delay, double maxVelScalingFactor)
{
    std::vector<double> arm_pose = {tf2Radians(angles[0]), 
                                    tf2Radians(angles[1]),
                                    tf2Radians(angles[2]),
                                    tf2Radians(angles[3])};
    group.setMaxVelocityScalingFactor(maxVelScalingFactor);
    group.setJointValueTarget(arm_pose);
    ROS_INFO("Moving single arm");
    group.move();
    ros::Duration(delay).sleep();
}

void moveBothArms(moveit::planning_interface::MoveGroupInterface& group, std::vector<double>& angles, double delay, double maxVelScalingFactor)
{
    std::vector<double> arm_pose = {tf2Radians(angles[0]), 
                                    tf2Radians(angles[1]),
                                    tf2Radians(angles[2]),
                                    tf2Radians(angles[3]),
                                    tf2Radians(angles[4]),
                                    tf2Radians(angles[5]),
                                    tf2Radians(angles[6]),
                                    tf2Radians(angles[7])};
    group.setMaxVelocityScalingFactor(maxVelScalingFactor);
    group.setJointValueTarget(arm_pose);
    ROS_INFO("Moving both arms");
    group.move();
    ros::Duration(delay).sleep();
}

void moveAllJoints(moveit::planning_interface::MoveGroupInterface& group, std::vector<double>& angles, double delay, double maxVelScalingFactor)
{
    std::vector<double> body_pose = {tf2Radians(angles[0]), 
                                    tf2Radians(angles[1]),
                                    tf2Radians(angles[2]),
                                    tf2Radians(angles[3]),
                                    tf2Radians(angles[4]),
                                    tf2Radians(angles[5]),
                                    tf2Radians(angles[6]),
                                    tf2Radians(angles[7]),
                                    tf2Radians(angles[8]),
                                    tf2Radians(angles[9]),
                                    tf2Radians(angles[10])};
    group.setMaxVelocityScalingFactor(maxVelScalingFactor);
    group.setJointValueTarget(body_pose);
    group.setPlanningTime(10);
    ROS_INFO("Moving all joints");
    group.move();
    ros::Duration(delay).sleep();
}

void moveGripper(moveit::planning_interface::MoveGroupInterface& group, double angle, double delay, double maxVelScalingFactor)
{
    std::vector<double> gripper_pose = {tf2Radians(angle), 
                                    tf2Radians(angle),
                                    tf2Radians(angle)};
    group.setMaxVelocityScalingFactor(maxVelScalingFactor);
    group.setJointValueTarget(gripper_pose);
    ROS_INFO("Moving head");
    group.move();
    ros::Duration(delay).sleep();
}

void inverted(std::vector<double>& angles, std::vector<double>& out_angles)
{
    for (auto &i : angles) out_angles.push_back(-i); 
}

void combine(std::vector<double>& l_angles, std::vector<double>& r_angles, std::vector<double>& out_angles)
{
    out_angles.insert(out_angles.begin(), r_angles.begin(), r_angles.end());
    out_angles.insert(out_angles.begin(), l_angles.begin(), l_angles.end());
}

int main(int argc, char **argv)
{

    ros::init(argc, argv, "move_group_joint_value_sample");

    ros::AsyncSpinner spinner(1);
    spinner.start();
    moveit_visual_tools::MoveItVisualTools visual_tools("torso_link_r4");

    // creating planning group
    moveit::planning_interface::MoveGroupInterface larm("left_arm");
    moveit::planning_interface::MoveGroupInterface rarm("right_arm");
    moveit::planning_interface::MoveGroupInterface barm("both_arms");
    moveit::planning_interface::MoveGroupInterface head("head");
    moveit::planning_interface::MoveGroupInterface all("whole_body");

    larm.setMaxVelocityScalingFactor(MAX_VEL_ARM);
    larm.setMaxAccelerationScalingFactor(MAX_ACC_ARM);

    // get current joint angles
    std::vector<double> l_arm_out = {-90, 29, -69, -34};
    std::vector<double> home = {0, 0, 0, 0};
    std::vector<double> b_home = {0, 0, 0, 0, 0, 0, 0, 0};
    std::vector<double> l_self = {-68, 12, -20, -77};
    std::vector<double> r_self;
    std::vector<double> r_arm_out;
    std::vector<double> b_arm_out;
    inverted(l_arm_out, r_arm_out);
    inverted(l_self, r_self);
    combine(l_arm_out, r_arm_out, b_arm_out);

    /***  pose for all joints ***************/
    std::vector<double> a_pose_0 = {0, 0, 0, 0,
                                    0, 0, 0,
                                    0, 0, 0, 0};
    std::vector<double> a_pose_1 = {-90, 29, -69, -34,
                                    -47, 7, -25,
                                    90, -29, 69, 34};
    std::vector<double> a_pose_2 = {-62, 44, -13, 97,
                                    35, -10, 5,
                                    63, -86, 46, 74};

    visual_tools.prompt("Press 'next' in the RvizVisualToolsGui window");
    int ctr = 0;

    while (ros::ok()) {
        printf("loop: %d\n", ctr);

        moveAllJoints(all, a_pose_0, 1.20, 0.5);
        moveAllJoints(all, a_pose_1, 1.20, 0.5);
        moveAllJoints(all, a_pose_2, 1.20, 0.5);

        ctr++;
    }
}

r4_control/include/r4_control/r4_hw_interface.h
#ifndef R4_HW_INTERFACE_H
#define R4_HW_INTERFACE_H

#include <ros_control_boilerplate/generic_hw_interface.h>
#include <r4_control/r4Cmd.h>
#include <r4_control/r4Sensor.h>
#include <control_msgs/FollowJointTrajectoryAction.h>
#include <std_msgs/Float32MultiArray.h>


const double DEG_TO_RAD = 0.017453292519943295769236907684886;
const double RAD_TO_DEG = 57.295779513082320876798154814105;

namespace r4_control_ns
{
/** \brief Hardware interface for a robot */
class R4HWInterface : public ros_control_boilerplate::GenericHWInterface
{
public:
  /**
   * \brief Constructor
   * \param nh - Node handle for topics.
   */
  R4HWInterface(ros::NodeHandle& nh, urdf::Model* urdf_model = NULL);

  /** \brief Initialize the robot hardware interface */
  virtual void init();

  /** \brief Read the state from the robot hardware. */
  virtual void read(ros::Duration& elapsed_time);

  /** \brief Write the command to the robot hardware. */
  virtual void write(ros::Duration& elapsed_time);

  /** \breif Enforce limits for all values before writing */
  virtual void enforceLimits(ros::Duration& period);

protected:
  ros::Subscriber sensor_sub;
  ros::Subscriber goal_pos_sub_l_arm;
  ros::Subscriber goal_pos_sub_r_arm;
  ros::Subscriber goal_pos_sub_head;
  void sensorCallback(const r4_control::r4Sensor::ConstPtr& msg);
  void lArmGoalCallback(const control_msgs::FollowJointTrajectoryActionGoal::ConstPtr& goal);
  void rArmGoalCallback(const control_msgs::FollowJointTrajectoryActionGoal::ConstPtr& goal);
  void headGoalCallback(const control_msgs::FollowJointTrajectoryActionGoal::ConstPtr& goal);


  ros::Publisher cmd_pub;
  std::vector<double> joint_position_prev_;
  std::vector<double> final_angle;
  double duration;
  // int interp_time = 0;
//   const float maxAngle[6] = {170.0,     0.0, 160.0, 120.0,  158.0, 60.0};
//   const float minAngle[6] = {-170.0, -135.0,   0.0, -75.0, -158.0, -15.0};


};  // class

}  // namespace ros_control_boilerplate

#endif


r4_control/src/r4_hw_interface.cpp
#include <r4_control/r4_hw_interface.h>
#include <control_msgs/FollowJointTrajectoryActionGoal.h>
#include <std_msgs/Float32MultiArray.h>

namespace r4_control_ns
{
    R4HWInterface::R4HWInterface(ros::NodeHandle &nh, urdf::Model *urdf_model)
        : ros_control_boilerplate::GenericHWInterface(nh, urdf_model)
    {
        sensor_sub = nh.subscribe("/r4/r4Sensor", 1, &R4HWInterface::sensorCallback, this);
        goal_pos_sub_l_arm = nh.subscribe("/l_arm_position_trajectory_controller/follow_joint_trajectory/goal", 1, &R4HWInterface::lArmGoalCallback, this);
        goal_pos_sub_r_arm = nh.subscribe("/r_arm_position_trajectory_controller/follow_joint_trajectory/goal", 1, &R4HWInterface::rArmGoalCallback, this);
        goal_pos_sub_head = nh.subscribe("/head_position_trajectory_controller/follow_joint_trajectory/goal", 1, &R4HWInterface::headGoalCallback, this);

        cmd_pub = nh.advertise<r4_control::r4Cmd>("/r4/r4Cmd", 3);
        ROS_INFO("R4HWInterface declared.");
    }

    void R4HWInterface::lArmGoalCallback(const control_msgs::FollowJointTrajectoryActionGoal::ConstPtr &msg)
    {
        int waypoints = msg->goal.trajectory.points.size();

        final_angle[0] = msg->goal.trajectory.points[waypoints - 1].positions[2];  // shoulder_pitch
        final_angle[1] = msg->goal.trajectory.points[waypoints - 1].positions[3];  // shoulder_roll
        final_angle[2] = msg->goal.trajectory.points[waypoints - 1].positions[1];  // elbow_yaw
        final_angle[3] = msg->goal.trajectory.points[waypoints - 1].positions[0];  // elbow_bend

        duration = msg->goal.trajectory.points[waypoints - 1].time_from_start.toSec();
    }

    void R4HWInterface::rArmGoalCallback(const control_msgs::FollowJointTrajectoryActionGoal::ConstPtr &msg)
    {
        int waypoints = msg->goal.trajectory.points.size();

        final_angle[4] = msg->goal.trajectory.points[waypoints - 1].positions[2];  // shoulder_pitch
        final_angle[5] = msg->goal.trajectory.points[waypoints - 1].positions[3];  // shoulder_roll
        final_angle[6] = msg->goal.trajectory.points[waypoints - 1].positions[1];  // elbow_yaw
        final_angle[7] = msg->goal.trajectory.points[waypoints - 1].positions[0];  // elbow_bend

        duration = msg->goal.trajectory.points[waypoints - 1].time_from_start.toSec();
    }

    // want: yaw roll pitch
    // got: pitch roll yaw
    void R4HWInterface::headGoalCallback(const control_msgs::FollowJointTrajectoryActionGoal::ConstPtr &msg)
    {
        int waypoints = msg->goal.trajectory.points.size();

        final_angle[8] = msg->goal.trajectory.points[waypoints - 1].positions[2];  //neck_yaw
        final_angle[9] = msg->goal.trajectory.points[waypoints - 1].positions[1];  //neck_roll
        final_angle[10] = msg->goal.trajectory.points[waypoints - 1].positions[0];  //neck_pitch

        duration = msg->goal.trajectory.points[waypoints - 1].time_from_start.toSec();
    }

    void R4HWInterface::sensorCallback(const r4_control::r4Sensor::ConstPtr &msg)
    {
        for (int i = 0; i < num_joints_; i++) {
            joint_position_[i] = msg->angle[i] * DEG_TO_RAD / 10.0; // angle received is multiplied by 10
        }
    }

    void R4HWInterface::init()
    {
        // Call parent class version of this function
        ros_control_boilerplate::GenericHWInterface::init();

        final_angle.resize(joint_position_.size());
        joint_position_prev_.resize(joint_position_.size());
        joint_position_command_.resize(joint_position_.size());

        ROS_INFO("R4HWInterface initiated.");
    }

    void R4HWInterface::read(ros::Duration &elapsed_time)
    {
        // No need to read since our write() command populates our state for us
        // ros::spinOnce();
    }

    void R4HWInterface::write(ros::Duration &elapsed_time)
    {
        static r4_control::r4Cmd joint_cmd;

        bool change_detected = false;
        for (int i = 0; i < num_joints_; i++) {
            if (joint_position_prev_[i] != joint_position_command_[i]) {
                change_detected = true;
                i = (int)num_joints_;
            }
        }

        if (change_detected) {
            for (int i = 0; i < num_joints_; i++) {
                joint_cmd.angle[i] = (int16_t)(final_angle[i] * RAD_TO_DEG * 10);
                joint_position_prev_[i] = joint_position_command_[i];
            }
            joint_cmd.time = (int16_t)(duration * 1000 / 16.667);
            cmd_pub.publish(joint_cmd);
        }
    }

    void R4HWInterface::enforceLimits(ros::Duration &period)
    {
        // Enforces position and velocity
        // pos_jnt_sat_interface_.enforceLimits(period);
    }

} // namespace r4_control_ns
